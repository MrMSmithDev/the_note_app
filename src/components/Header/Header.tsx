import Modal from '@components/Modal/Modal';
import SignInForm from '@components/SignInForm';
import SignUpForm from '@components/SignUpForm';
import ThemeToggle from '@components/ThemeToggle';
import React, { useRef, useState } from 'react';

const Header: React.FC = () => {
  const [signInFormActive, setSignInFormActive] = useState(false);
  const [signUpFormActive, setSignUpFormActive] = useState(false);

  const ref = useRef(null);

  function toggleForm() {
    setSignInFormActive((prev) => !prev);
    setSignUpFormActive((prev) => !prev);
  }

  return (
    <header className="bg-background dark:bg-background-dark sticky top-0 flex px-3 md:px-4 shadow-xl dark:shadow-lg dark:shadow-gray-700 z-10">
      <p className="font-kanit italic font-bold text-2xl text-transparent bg-clip-text mr-auto bg-linear-to-tr from-blue-400 to-blue-600 py-2 select-none">
        NOTES
      </p>
      <div className="ml-auto my-auto">
        {signInFormActive && !signUpFormActive && (
          <Modal closeModal={() => setSignInFormActive(false)} btnRef={ref}>
            <SignInForm
              closeForm={() => setSignInFormActive(false)}
              toggleForm={toggleForm}
            />
          </Modal>
        )}
        {signUpFormActive && !signInFormActive && (
          <Modal closeModal={() => setSignInFormActive(false)} btnRef={ref}>
            <SignUpForm
              closeForm={() => setSignInFormActive(false)}
              toggleForm={toggleForm}
            />
          </Modal>
        )}
        <button
          ref={ref}
          className="bg-blue-500 py-1 px-4 rounded-md font-quicksand font-bold text-gray-100 cursor-pointer hover:animate-jello"
          onClick={() => setSignInFormActive(true)}
        >
          Log in
        </button>
      </div>
      <ThemeToggle />
    </header>
  );
};

export default Header;
