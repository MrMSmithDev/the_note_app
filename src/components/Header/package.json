{
    "name": "header",
    "version": "1.0.0",
    "main": "Header.tsx",
    "dependencies": {
      "react": "^19.0.0"
    }
}