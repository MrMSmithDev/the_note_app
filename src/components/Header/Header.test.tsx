import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Header from './Header';

jest.mock('@components/Modal', () => require('@__mocks__/Modal'));
jest.mock('@components/ThemeToggle', () => require('@__mocks__/ThemeToggle'));
jest.mock('@components/SignInForm', () => require('@__mocks__/SignInForm'));
jest.mock('@components/SignUpForm', () => require('@__mocks__/SignUpForm'));

describe('Header component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the site name', () => {
    render(<Header />);

    expect(screen.getByText('NOTES')).toBeInTheDocument();
  });

  it('renders the log in button', () => {
    render(<Header />);

    expect(screen.getByText('Log in')).toBeInTheDocument();
  });

  it('renders the theme toggle component', () => {
    render(<Header />);

    expect(screen.getByTestId('mock-theme-toggle')).toBeInTheDocument();
  });

  it('opens the SignInForm when clicking the log in button', () => {
    render(<Header />);

    fireEvent.click(screen.getByText('Log in'));

    expect(screen.getByTestId('mock-modal')).toBeInTheDocument();
    expect(screen.getByTestId('mock-sign-in-form')).toBeInTheDocument();
  });

  it('closes the SignInForm when clicking close', async () => {
    render(<Header />);

    fireEvent.click(screen.getByText('Log in'));
    await waitFor(() => {
      expect(screen.getByTestId('mock-sign-in-form')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('close sign in'));
    await waitFor(() => {
      expect(screen.queryByTestId('mock-sign-in-form')).toBeNull();
    });
  });

  it('switches from SignInForm to SignUpForm when clicking toggle', async () => {
    render(<Header />);

    fireEvent.click(screen.getByText('Log in'));
    await waitFor(() => {
      expect(screen.getByTestId('mock-sign-in-form')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('switch to sign up form'));
    await waitFor(() => {
      expect(screen.getByTestId('mock-sign-up-form')).toBeInTheDocument();
    });
  });

  it('closes the SignUpForm when clicking close', async () => {
    render(<Header />);

    fireEvent.click(screen.getByText('Log in'));
    await waitFor(() => {
      expect(screen.getByTestId('mock-sign-in-form')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('switch to sign up form'));
    await waitFor(() => {
      expect(screen.getByTestId('mock-sign-up-form')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('close sign up'));
    await waitFor(() => {
      expect(screen.queryByTestId('mock-sign-up-form')).toBeNull();
    });
  });
});
