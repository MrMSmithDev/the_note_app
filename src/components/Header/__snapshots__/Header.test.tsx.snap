// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders to match snapshot 1`] = `
<DocumentFragment>
  <header
    class="bg-background dark:bg-background-dark sticky top-0 flex px-3 md:px-4 shadow-xl dark:shadow-lg dark:shadow-gray-700 z-10"
  >
    <p
      class="font-kanit italic font-bold text-2xl text-transparent bg-clip-text mr-auto bg-linear-to-tr from-blue-400 to-blue-600 py-2 select-none"
    >
      NOTES
    </p>
    <div
      class="ml-auto my-auto"
    >
      <button
        class="bg-blue-500 py-1 px-4 rounded-md font-quicksand font-bold text-gray-100 cursor-pointer hover:animate-jello"
      >
        Log in
      </button>
    </div>
    <div
      data-testid="mock-theme-toggle"
    />
  </header>
</DocumentFragment>
`;
